// @yext/analytics/lib/commonjs/models/pages/events/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhoneCallEvent = exports.PageViewEvent = exports.DrivingDirectionEvent = exports.WebsiteEvent = exports.CtaClick = exports.DefaultPagesEventNames = void 0;
var DefaultPagesEventNames_1 = require("./DefaultPagesEventNames");
Object.defineProperty(exports, "DefaultPagesEventNames", { enumerable: true, get: function () { return DefaultPagesEventNames_1.DefaultPagesEventNames; } });
var CtaClick_1 = require("./CtaClick");
Object.defineProperty(exports, "CtaClick", { enumerable: true, get: function () { return CtaClick_1.CtaClick; } });
var WebsiteEvent_1 = require("./WebsiteEvent");
Object.defineProperty(exports, "WebsiteEvent", { enumerable: true, get: function () { return WebsiteEvent_1.WebsiteEvent; } });
var DrivingDirectionEvent_1 = require("./DrivingDirectionEvent");
Object.defineProperty(exports, "DrivingDirectionEvent", { enumerable: true, get: function () { return DrivingDirectionEvent_1.DrivingDirectionEvent; } });
var PageViewEvent_1 = require("./PageViewEvent");
Object.defineProperty(exports, "PageViewEvent", { enumerable: true, get: function () { return PageViewEvent_1.PageViewEvent; } });
var PhoneCallEvent_1 = require("./PhoneCallEvent");
Object.defineProperty(exports, "PhoneCallEvent", { enumerable: true, get: function () { return PhoneCallEvent_1.PhoneCallEvent; } });
