// @yext/search-core/dist/commonjs/src/transformers/searchservice/createUniversalSearchResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createUniversalSearchResponse = void 0;
const createVerticalResults_1 = require("./createVerticalResults");
const createDirectAnswer_1 = require("./createDirectAnswer");
const createSpellCheck_1 = require("./createSpellCheck");
const createLocationBias_1 = require("./createLocationBias");
const createFailedVertical_1 = require("./createFailedVertical");
function createUniversalSearchResponse(data) {
    var _a;
    const verticalResults = Array.isArray(data.response.modules)
        ? data.response.modules.map(createVerticalResults_1.createVerticalResults)
        : [];
    return {
        verticalResults: verticalResults,
        queryId: data.response.queryId,
        directAnswer: data.response.directAnswer && (0, createDirectAnswer_1.createDirectAnswer)(data.response.directAnswer),
        searchIntents: data.response.searchIntents,
        spellCheck: data.response.spellCheck && (0, createSpellCheck_1.createSpellCheck)(data.response.spellCheck),
        locationBias: data.response.locationBias && (0, createLocationBias_1.createLocationBias)(data.response.locationBias),
        uuid: data.meta.uuid,
        queryRulesActionsData: data.response.queryRulesActionsData,
        failedVerticals: (_a = data.response.failedVerticals) === null || _a === void 0 ? void 0 : _a.map(createFailedVertical_1.createFailedVertical)
    };
}
exports.createUniversalSearchResponse = createUniversalSearchResponse;
