// @yext/search-core/dist/commonjs/src/models/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Matcher = exports.ErrorType = exports.SpellCheckType = exports.Source = exports.SearchIntent = exports.LocationBiasMethod = exports.BuiltInFieldType = exports.DirectAnswerType = exports.AppliedQueryFilterType = exports.Direction = exports.SortType = exports.QueryTrigger = exports.QuerySource = exports.FilterCombinator = exports.SearchError = void 0;
// Search API models
var SearchError_1 = require("./searchapi/SearchError");
Object.defineProperty(exports, "SearchError", { enumerable: true, get: function () { return SearchError_1.SearchError; } });
// Autocomplete service
__exportStar(require("./autocompleteservice/AutocompleteRequest"), exports);
__exportStar(require("./autocompleteservice/AutocompleteResponse"), exports);
// Search service request models
var StaticFilter_1 = require("./searchservice/request/StaticFilter");
Object.defineProperty(exports, "FilterCombinator", { enumerable: true, get: function () { return StaticFilter_1.FilterCombinator; } });
var QuerySource_1 = require("./searchservice/request/QuerySource");
Object.defineProperty(exports, "QuerySource", { enumerable: true, get: function () { return QuerySource_1.QuerySource; } });
var QueryTrigger_1 = require("./searchservice/request/QueryTrigger");
Object.defineProperty(exports, "QueryTrigger", { enumerable: true, get: function () { return QueryTrigger_1.QueryTrigger; } });
var SortType_1 = require("./searchservice/request/SortType");
Object.defineProperty(exports, "SortType", { enumerable: true, get: function () { return SortType_1.SortType; } });
var Direction_1 = require("./searchservice/request/Direction");
Object.defineProperty(exports, "Direction", { enumerable: true, get: function () { return Direction_1.Direction; } });
var AppliedQueryFilterType_1 = require("./searchservice/response/AppliedQueryFilterType");
Object.defineProperty(exports, "AppliedQueryFilterType", { enumerable: true, get: function () { return AppliedQueryFilterType_1.AppliedQueryFilterType; } });
var DirectAnswerType_1 = require("./searchservice/response/directanswer/DirectAnswerType");
Object.defineProperty(exports, "DirectAnswerType", { enumerable: true, get: function () { return DirectAnswerType_1.DirectAnswerType; } });
var BuiltInFieldType_1 = require("./searchservice/response/directanswer/BuiltInFieldType");
Object.defineProperty(exports, "BuiltInFieldType", { enumerable: true, get: function () { return BuiltInFieldType_1.BuiltInFieldType; } });
__exportStar(require("./searchservice/response/directanswer/fieldvaluedirectanswers"), exports);
__exportStar(require("./searchservice/response/directanswer/featuredsnippetdirectanswers"), exports);
var LocationBias_1 = require("./searchservice/response/LocationBias");
Object.defineProperty(exports, "LocationBiasMethod", { enumerable: true, get: function () { return LocationBias_1.LocationBiasMethod; } });
var SearchIntent_1 = require("./searchservice/response/SearchIntent");
Object.defineProperty(exports, "SearchIntent", { enumerable: true, get: function () { return SearchIntent_1.SearchIntent; } });
var Source_1 = require("./searchservice/response/Source");
Object.defineProperty(exports, "Source", { enumerable: true, get: function () { return Source_1.Source; } });
var SpellCheck_1 = require("./searchservice/response/SpellCheck");
Object.defineProperty(exports, "SpellCheckType", { enumerable: true, get: function () { return SpellCheck_1.SpellCheckType; } });
var ErrorType_1 = require("./searchservice/response/ErrorType");
Object.defineProperty(exports, "ErrorType", { enumerable: true, get: function () { return ErrorType_1.ErrorType; } });
// Search service common models
var Matcher_1 = require("./searchservice/common/Matcher");
Object.defineProperty(exports, "Matcher", { enumerable: true, get: function () { return Matcher_1.Matcher; } });
