// @yext/search-core/dist/esm/src/transformers/autocompleteservice/createAutocompleteResponse.js
import { SearchError } from '../../models/searchapi/SearchError';
import { createAutocompleteResult } from './createAutocompleteResult';
export function createAutocompleteResponse(data) {
    if (!data.response) {
        throw new Error('The autocomplete data does not contain a response property');
    }
    if (!Object.keys(data.response).length) {
        throw new Error('The autocomplete response is empty');
    }
    var response = data.response;
    var responseResults = response.results.map(createAutocompleteResult);
    var inputIntents = response.input ? response.input.queryIntents : [];
    return {
        results: responseResults,
        queryId: response.queryId,
        inputIntents: inputIntents || [],
        uuid: data.meta.uuid
    };
}
export function createFilterSearchResponse(data) {
    if (!data.response) {
        throw new Error('The autocomplete data does not contain a response property');
    }
    if (!Object.keys(data.response).length) {
        throw new Error('The autocomplete response is empty');
    }
    var response = data.response;
    if (response.failedVerticals && response.failedVerticals.length != 0) {
        var error = response.failedVerticals[0];
        throw new SearchError(error.details.description, error.details.responseCode, error.errorType);
    }
    var sections = response.sections.map(function (section) { return ({
        label: section.label,
        results: section.results.map(createAutocompleteResult)
    }); });
    return {
        sections: sections,
        queryId: response.queryId,
        businessId: response.businessId,
        uuid: data.meta.uuid
    };
}
