// @yext/pages/dist/components/link/link.stories.js
import React from "react";
import { Link } from "./link.js";
const meta = {
  title: "components/Link",
  component: Link
};
var link_stories_default = meta;
const Template = (args) => /* @__PURE__ */ React.createElement(Link, {
  ...args
});
const HREF = Template.bind({});
HREF.args = {
  href: "https://yext.com",
  children: "Learn More"
};
const CTA = Template.bind({});
CTA.args = {
  cta: {
    link: "https://yext.com",
    label: "Learn More",
    linkType: "URL"
  }
};
const Phone = Template.bind({});
Phone.args = {
  cta: {
    link: "2023013404",
    label: "Call us",
    linkType: "Phone"
  }
};
const Email = Template.bind({});
Email.args = {
  cta: {
    link: "consulting-dev@yext.com",
    label: "Email Me",
    linkType: "Email"
  }
};
const Detect_Email = Template.bind({});
Detect_Email.args = {
  cta: {
    link: "consulting-dev@yext.com"
  }
};
const CTA_With_Children = Template.bind({});
CTA_With_Children.args = {
  cta: {
    link: "https://yext.com",
    label: "Learn More",
    linkType: "URL"
  },
  children: "Override label"
};
export {
  CTA,
  CTA_With_Children,
  Detect_Email,
  Email,
  HREF,
  Phone,
  link_stories_default as default
};
