// @yext/pages/dist/dev/server/server.js
import express from "express";
import { createServer as createViteServer } from "vite";
import { serverRenderRoute } from "./middleware/serverRenderRoute.js";
import { ignoreFavicon } from "./middleware/ignoreFavicon.js";
import { errorMiddleware } from "./middleware/errorMiddleware.js";
import { devServerPort } from "./middleware/constants.js";
import { indexPage } from "./middleware/indexPage.js";
import { generateTestData } from "./ssr/generateTestData.js";
import { ProjectStructure } from "../../common/src/project/structure.js";
import { finalSlashRedirect } from "./middleware/finalSlashRedirect.js";
import { serverRenderSlugRoute } from "./middleware/serverRenderSlugRoute.js";
import { processEnvVariables } from "../../util/processEnvVariables.js";
const createServer = async (dynamicGenerateData, useProdURLs, scope) => {
  const app = express();
  const projectStructure = new ProjectStructure({
    filepathsConfig: {
      scope
    }
  });
  const vite = await createViteServer({
    server: {
      middlewareMode: true
    },
    appType: "custom",
    envDir: projectStructure.envVarDir,
    envPrefix: projectStructure.envVarPrefix,
    define: processEnvVariables(projectStructure.envVarPrefix)
  });
  app.use(vite.middlewares);
  app.use(ignoreFavicon);
  app.use(finalSlashRedirect);
  let displayGenerateTestDataWarning = false;
  if (dynamicGenerateData) {
    displayGenerateTestDataWarning = !await generateTestData(
      projectStructure.scope
    );
  }
  app.use(
    /^\/(.+)/,
    useProdURLs ? serverRenderSlugRoute({ vite, dynamicGenerateData, projectStructure }) : serverRenderRoute({ vite, dynamicGenerateData, projectStructure })
  );
  app.use(
    "/",
    indexPage({
      vite,
      dynamicGenerateData,
      displayGenerateTestDataWarning,
      useProdURLs,
      projectStructure
    })
  );
  app.use(errorMiddleware(vite));
  app.listen(
    devServerPort,
    () => process.stdout.write(`listening on :${devServerPort}
`)
  );
};
export {
  createServer
};
