// @yext/pages/dist/common/src/template/internal/loader.js
import {
  convertTemplateModuleToTemplateModuleInternal
} from "./types.js";
import esbuild from "esbuild";
import { importFromString } from "module-from-string";
import { pathToFileURL } from "url";
const TEMP_DIR = ".temp";
const loadTemplateModules = async (templateModulePaths, transpile, adjustForFingerprintedAsset) => {
  const importedModules = [];
  for (const templateModulePath of templateModulePaths) {
    let templateModule = {};
    try {
      if (transpile) {
        const buildResult = await esbuild.build({
          entryPoints: [templateModulePath],
          outdir: TEMP_DIR,
          write: false,
          format: "esm",
          bundle: true
        });
        templateModule = await importFromString(
          buildResult.outputFiles[0].text
        );
      } else {
        templateModule = await import(pathToFileURL(templateModulePath).toString());
      }
    } catch (e) {
      throw new Error(`Could not import ${templateModulePath} ${e}`);
    }
    const templateModuleInternal = convertTemplateModuleToTemplateModuleInternal(
      templateModulePath,
      templateModule,
      adjustForFingerprintedAsset
    );
    importedModules.push({
      ...templateModuleInternal,
      path: templateModulePath
    });
  }
  return importedModules.reduce((prev, module) => {
    if (prev.has(module.config.name)) {
      throw new Error(
        `Templates must have unique feature names. Found multiple modules with "${module.config.name}"`
      );
    }
    return prev.set(module.config.name, module);
  }, /* @__PURE__ */ new Map());
};
export {
  loadTemplateModules
};
