// @yext/analytics/lib/esm/models/search/SearchAnalyticsEventType.js
/**
 * Represents the possible analytics event types.
 *
 * @public
 */
export var SearchAnalyticsEventType;
(function (SearchAnalyticsEventType) {
    /** {@link AllTabNavigationEvent} */
    SearchAnalyticsEventType["AllTabNavigation"] = "ALL_TAB_NAVIGATION";
    /** {@link AllTabNavigationEvent} */
    SearchAnalyticsEventType["VerticalTabNavigation"] = "VERTICAL_TAB_NAVIGATION";
    /** {@link ScrollEvent} */
    SearchAnalyticsEventType["ScrollToBottomOfPage"] = "SCROLL_TO_BOTTOM_OF_PAGE";
    /** {@link QuestionSubmissionEvent} */
    SearchAnalyticsEventType["QuestionFocus"] = "QUESTION_FOCUS";
    /** {@link QuestionSubmissionEvent} */
    SearchAnalyticsEventType["QuestionSubmit"] = "QUESTION_SUBMIT";
    /** {@link AccordionToggleEvent} */
    SearchAnalyticsEventType["RowExpand"] = "ROW_EXPAND";
    /** {@link AccordionToggleEvent} */
    SearchAnalyticsEventType["RowCollapse"] = "ROW_COLLAPSE";
    /** {@link ThumbsFeedbackEvent} */
    SearchAnalyticsEventType["ThumbsUp"] = "THUMBS_UP";
    /** {@link ThumbsFeedbackEvent} */
    SearchAnalyticsEventType["ThumbsDown"] = "THUMBS_DOWN";
    /** {@link VoiceSearchEvent} */
    SearchAnalyticsEventType["VoiceStart"] = "VOICE_START";
    /** {@link VoiceSearchEvent} */
    SearchAnalyticsEventType["VoiceStop"] = "VOICE_STOP";
    /** {@link SearchBarImpressionEvent} */
    SearchAnalyticsEventType["SearchBarImpression"] = "SEARCH_BAR_IMPRESSION";
    /** {@link SearchClearEvent} */
    SearchAnalyticsEventType["SearchClearButton"] = "SEARCH_CLEAR_BUTTON";
    /** {@link PaginationEvent} */
    SearchAnalyticsEventType["Paginate"] = "PAGINATE";
    /** {@link AutocompleteEvent} */
    SearchAnalyticsEventType["AutocompleteSelection"] = "AUTO_COMPLETE_SELECTION";
    /** {@link VerticalViewAllEvent} */
    SearchAnalyticsEventType["VerticalViewAll"] = "VERTICAL_VIEW_ALL";
    // CTA event types
    /** {@link CtaEvent} */
    SearchAnalyticsEventType["CtaClick"] = "CTA_CLICK";
    /** {@link CtaEvent} */
    SearchAnalyticsEventType["TitleClick"] = "TITLE_CLICK";
    /** {@link CtaEvent} */
    SearchAnalyticsEventType["TapToCall"] = "TAP_TO_CALL";
    /** {@link CtaEvent} */
    SearchAnalyticsEventType["OrderNow"] = "ORDER_NOW";
    /** {@link CtaEvent} */
    SearchAnalyticsEventType["AddToCart"] = "ADD_TO_CART";
    /** {@link CtaEvent} */
    SearchAnalyticsEventType["ApplyNow"] = "APPLY_NOW";
    /** {@link CtaEvent} */
    SearchAnalyticsEventType["DrivingDirections"] = "DRIVING_DIRECTIONS";
    /** {@link CtaEvent} */
    SearchAnalyticsEventType["ViewWebsite"] = "VIEW_WEBSITE";
    /** {@link CtaEvent} */
    SearchAnalyticsEventType["Email"] = "EMAIL";
    /** {@link CtaEvent} */
    SearchAnalyticsEventType["BookAppointment"] = "BOOK_APPOINTMENT";
    /** {@link CtaEvent} */
    SearchAnalyticsEventType["Rsvp"] = "RSVP";
    // Search duration event types
    /** {@link SearchDurationEvent} */
    SearchAnalyticsEventType["ResultsHidden"] = "RESULTS_HIDDEN";
    /** {@link SearchDurationEvent} */
    SearchAnalyticsEventType["ResultsUnhidden"] = "RESULTS_UNHIDDEN";
    /** {@link SearchDurationEvent} */
    SearchAnalyticsEventType["FollowUpQuery"] = "FOLLOW_UP_QUERY";
})(SearchAnalyticsEventType || (SearchAnalyticsEventType = {}));
export { SearchAnalyticsEventType as AnalyticsEventType };
