// @yext/studio/lib/src/components/AddElementMenu/AddElementMenu.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState } from "react";
import AddElementsList from "./AddElementsList";
import { useCallback } from "react";
import renderIconForType from "../common/renderIconForType";
import useStudioStore from "../../store/useStudioStore";
import { FileMetadataKind } from "@yext/studio-plugin";
export var ElementType;
(function (ElementType) {
    ElementType["Components"] = "Components";
    ElementType["Containers"] = "Containers";
    ElementType["Modules"] = "Modules";
})(ElementType || (ElementType = {}));
/**
 * A menu for adding elements to the page.
 */
export default function AddElementMenu() {
    const [activeType, setType] = useState(ElementType.Components);
    return (_jsxs("div", { className: "absolute z-20 rounded bg-white text-sm text-gray-700 shadow-lg", children: [_jsx(ElementTypeSwitcher, { activeType: activeType, setType: setType }), _jsx(AddElementsList, { activeType: activeType })] }));
}
function ElementTypeSwitcher(props) {
    const { activeType, setType } = props;
    const UUIDToFileMetadata = useStudioStore((store) => {
        return store.fileMetadatas.UUIDToFileMetadata;
    });
    const numAvailableModules = Object.values(UUIDToFileMetadata).filter((metadata) => metadata.kind === FileMetadataKind.Module).length;
    return (_jsx("div", { className: "flex px-4 pt-2 border-b", children: Object.values(ElementType).map((elementType) => {
            if (elementType === ElementType.Modules && numAvailableModules === 0) {
                return null;
            }
            return (_jsx(ElementTypeButton, { elementType: elementType, isActiveType: elementType === activeType, handleClick: setType }, elementType));
        }) }));
}
function ElementTypeButton(props) {
    const { isActiveType, elementType, handleClick } = props;
    const onClick = useCallback(() => {
        handleClick(ElementType[elementType]);
    }, [elementType, handleClick]);
    return (_jsxs("button", { className: "p-2 mx-2 flex items-center border-b-2 border-transparent disabled:border-blue-600", onClick: onClick, disabled: isActiveType, children: [_jsx("span", { className: "mr-2 pt-0.5", children: renderIconForType(elementType) }), _jsx("span", { children: elementType })] }));
}
