// @yext/pages/dist/generate/features/features.js
import path from "path";
import { Path } from "../../common/src/project/path.js";
import {
  defaultProjectStructureConfig
} from "../../common/src/project/structure.js";
import { getTemplateFilepaths } from "../../common/src/template/internal/getTemplateFilepaths.js";
import { loadTemplateModules } from "../../common/src/template/internal/loader.js";
import { createFeaturesJson } from "./createFeaturesJson.js";
const handler = async ({ scope }) => {
  const templatesRoot = defaultProjectStructureConfig.filepathsConfig.templatesRoot;
  const sitesConfigRoot = defaultProjectStructureConfig.filepathsConfig.sitesConfigRoot;
  const templateRootAbsolutePath = path.join(process.cwd(), templatesRoot);
  const templateFilepaths = getTemplateFilepaths(
    scope ? [
      new Path(path.join(templateRootAbsolutePath, scope)),
      new Path(templateRootAbsolutePath)
    ] : [new Path(templateRootAbsolutePath)]
  );
  const templateModules = await loadTemplateModules(
    templateFilepaths,
    true,
    false
  );
  const featuresFilepath = path.join(
    process.cwd(),
    sitesConfigRoot,
    scope ?? "",
    "features.json"
  );
  await createFeaturesJson(templateModules, featuresFilepath);
};
const featureCommandModule = {
  command: "features",
  describe: "Generates features.json file",
  builder: (yargs) => {
    return yargs.option("scope", {
      describe: "The subfolder to scope the served templates from",
      type: "string",
      demandOption: false
    });
  },
  handler
};
export {
  featureCommandModule
};
