// @yext/studio/lib/src/store/StudioActions/RepeaterActions.js
import { ComponentStateKind, } from "@yext/studio-plugin";
export default class RepeaterActions {
    studioActions;
    constructor(studioActions) {
        this.studioActions = studioActions;
    }
    /**
     * Turns the standard component or module into a repeater.
     */
    addRepeater = (componentState) => {
        const { uuid, parentUUID, ...remainingComponentState } = componentState;
        const newComponentState = {
            kind: ComponentStateKind.Repeater,
            uuid,
            parentUUID,
            repeatedComponent: remainingComponentState,
            listExpression: "",
        };
        this.studioActions.replaceComponentState(componentState.uuid, newComponentState);
    };
    /**
     * Turns the repeater component back into a regular component or module.
     */
    removeRepeater = (componentState) => {
        const newComponentState = {
            ...componentState.repeatedComponent,
            uuid: componentState.uuid,
            parentUUID: componentState.parentUUID,
        };
        this.studioActions.replaceComponentState(componentState.uuid, newComponentState);
    };
}
