// @yext/pages/dist/components/address/address.stories.js
import React from "react";
import { Address as AddressComponent } from ".";
import {
  ADDRESS,
  Arlington,
  Berlin,
  Chicago,
  London,
  Miami,
  NewYork,
  Paris,
  SanFrancisco
} from "./sampleData.js";
const meta = {
  title: "components/Address",
  component: AddressComponent
};
var address_stories_default = meta;
const Template = (args) => /* @__PURE__ */ React.createElement(AddressComponent, {
  ...args
});
const Address = Template.bind({});
Address.args = {
  address: ADDRESS
};
const Address_Arlington = Template.bind({});
Address_Arlington.args = {
  address: Arlington
};
const Address_Berlin = Template.bind({});
Address_Berlin.args = {
  address: Berlin
};
const Address_Chicago = Template.bind({});
Address_Chicago.args = {
  address: Chicago
};
const Address_London = Template.bind({});
Address_London.args = {
  address: London
};
const Address_Miami = Template.bind({});
Address_Miami.args = {
  address: Miami
};
const Address_NewYork = Template.bind({});
Address_NewYork.args = {
  address: NewYork
};
const Address_Paris = Template.bind({});
Address_Paris.args = {
  address: Paris
};
const Address_SanFrancisco = Template.bind({});
Address_SanFrancisco.args = {
  address: SanFrancisco
};
export {
  Address,
  Address_Arlington,
  Address_Berlin,
  Address_Chicago,
  Address_London,
  Address_Miami,
  Address_NewYork,
  Address_Paris,
  Address_SanFrancisco,
  address_stories_default as default
};
