// @yext/pages/dist/components/address/sampleData.js
const ADDRESS = {
  city: "Birmingham",
  countryCode: "US",
  line1: "1716 University Boulevard",
  localizedCountryName: "United States",
  localizedRegionName: "Alabama",
  postalCode: "35294",
  region: "AL"
};
const Arlington = {
  city: "Arlington",
  countryCode: "US",
  line1: "1101 Wilson Blvd",
  localizedCountryName: "United States",
  localizedRegionName: "Virginia",
  postalCode: "22201",
  region: "VA"
};
const Berlin = {
  city: "Berlin",
  countryCode: "DE",
  line1: "Eichhornstra\xDFe 3",
  line2: "c/o WeWork Atrium Tower",
  localizedCountryName: "Germany",
  postalCode: "10785"
};
const Chicago = {
  city: "Chicago",
  countryCode: "US",
  line1: "500 West Madison Street",
  line2: "Suite 3160",
  localizedCountryName: "United States",
  localizedRegionName: "Illinois",
  postalCode: "60661"
};
const London = {
  city: "London",
  countryCode: "GB",
  line1: "2 Great Titchfield St",
  line2: "7th Floor, United Kingdom House",
  localizedCountryName: "United Kingdon",
  localizedRegionName: "London",
  postalCode: "W1D 1NN",
  region: "GB-LND"
};
const Miami = {
  city: "Miami",
  countryCode: "US",
  line1: "600 Brickell Ave",
  localizedCountryName: "United States",
  localizedRegionName: "Florida",
  postalCode: "33131",
  region: "FL"
};
const Milan = {
  city: "Milan",
  countryCode: "IT",
  line1: "Via San Raffaele 1",
  localizedCountryName: "Italy",
  postalCode: "20121",
  region: "MI"
};
const NewYork = {
  city: "New York",
  countryCode: "US",
  line1: "61 9th Ave",
  localizedCountryName: "United States",
  localizedRegionName: "New York",
  postalCode: "10011",
  region: "NY"
};
const Paris = {
  city: "Paris",
  countryCode: "FR",
  line1: "83 Avenue de la Grande Arm\xE9e",
  localizedCountryName: "France",
  postalCode: "75016"
};
const SanFrancisco = {
  city: "San Francisco",
  countryCode: "US",
  line1: "535 Mission Street",
  line2: "14th Floor",
  localizedCountryName: "United States",
  localizedRegionName: "California",
  postalCode: "94105",
  region: "CA"
};
export {
  ADDRESS,
  Arlington,
  Berlin,
  Chicago,
  London,
  Miami,
  Milan,
  NewYork,
  Paris,
  SanFrancisco
};
