// @yext/analytics/lib/commonjs/infra/HttpRequester.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpRequester = void 0;
const cross_fetch_1 = __importDefault(require("cross-fetch"));
/**
 * Responsible for making web requests.
 *
 */
class HttpRequester {
    post(url, body) {
        const data = JSON.stringify(body);
        const fetchInit = {
            method: 'POST',
            body: data,
            keepalive: true
        };
        if (typeof (window) !== 'undefined' && window.fetch) {
            return window.fetch(url, fetchInit);
        }
        return (0, cross_fetch_1.default)(url, fetchInit);
    }
    get(url) {
        const fetchInit = {
            method: 'GET',
            mode: 'no-cors',
        };
        if (typeof (window) !== 'undefined' && window.fetch) {
            return window.fetch(url, fetchInit);
        }
        return (0, cross_fetch_1.default)(url, fetchInit);
    }
}
exports.HttpRequester = HttpRequester;
