// @dhmk/utils/lib/std.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.flatMap = exports.isPlainObject = exports.assert = exports.namespace = exports.join = exports.mergeDeep = exports.merge = exports.deepReadonly = exports.call = exports.createFactoryAs = exports.createFactory = exports.getter = exports.objectTransform = exports.objectMap = exports.objectFrom = exports.shallowEqual = exports.arraysEqual = exports.range = exports.id = exports.noop = void 0;
var noop = function () { };
exports.noop = noop;
var id = function (x) { return x; };
exports.id = id;
var range = function (count, start) {
    if (start === void 0) { start = 0; }
    return Array.from(Array(count).keys()).map(function (i) { return start + i; });
};
exports.range = range;
var arraysEqual = function (a, b) {
    return a.length === b.length && a.every(function (av, i) { return av === b[i]; });
};
exports.arraysEqual = arraysEqual;
var shallowEqual = function (a, b) {
    var ka = Object.keys(a);
    var kb = Object.keys(b);
    return ka.length === kb.length && ka.every(function (k) { return a[k] === b[k]; });
};
exports.shallowEqual = shallowEqual;
function objectFrom(entries) {
    return Array.from(entries).reduce(function (res, _a) {
        var k = _a[0], v = _a[1];
        res[k] = v;
        return res;
    }, {});
}
exports.objectFrom = objectFrom;
function objectMap(src, mapValue, mapKey) {
    if (mapKey === void 0) { mapKey = exports.id; }
    return objectFrom(Object.keys(src).map(function (k) { return [
        mapKey(k, src[k], src),
        mapValue(src[k], k, src),
    ]; }));
}
exports.objectMap = objectMap;
function objectTransform(src, mapEntry) {
    return objectFrom(Object.keys(src)
        .map(function (k) { return mapEntry(src[k], k, src); })
        .filter(Boolean));
}
exports.objectTransform = objectTransform;
function getter(t, k, value) {
    if (arguments.length === 2) {
        return { use: function (value) { return getter(t, k, value); } };
    }
    else {
        Object.defineProperty(t, k, { value: value, enumerable: true });
        return value;
    }
}
exports.getter = getter;
function createFactory(c) {
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        // @ts-ignore
        return new (c.bind.apply(c, __spreadArray([void 0], args, false)))();
    };
}
exports.createFactory = createFactory;
function createFactoryAs() {
    return function (c) {
        return function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            // @ts-ignore
            return new (c.bind.apply(c, __spreadArray([void 0], args, false)))();
        };
    };
}
exports.createFactoryAs = createFactoryAs;
var call = function (fn) { return fn(); };
exports.call = call;
var deepReadonly = function (x) { return x; };
exports.deepReadonly = deepReadonly;
var merge = function (a, b) {
    return b ? __assign(__assign({}, a), b) : function (b) { return (__assign(__assign({}, b), a)); };
};
exports.merge = merge;
var mergeDeepRec = function (a, b) {
    if (!(0, exports.isPlainObject)(a) || !(0, exports.isPlainObject)(b))
        return b;
    var res = __assign({}, a);
    for (var k in b) {
        var v = b[k];
        res[k] = mergeDeepRec(a[k], v);
    }
    return res;
};
var mergeDeep = function (a, b) {
    return b ? mergeDeepRec(a, b) : function (b) { return mergeDeepRec(b, a); };
};
exports.mergeDeep = mergeDeep;
var join = function (a, sep, b) { return (a + sep + b); };
exports.join = join;
var namespace = function (s) {
    return function (suffix) {
        return (0, exports.join)(s, "/", suffix);
    };
};
exports.namespace = namespace;
function assert(condition, msg) {
    if (msg === void 0) { msg = "Assertion failed"; }
    if (!condition)
        throw new Error(msg);
}
exports.assert = assert;
var isPlainObject = function (x) {
    return !!x && typeof x === "object" && Object.getPrototypeOf(x) === Object.prototype;
};
exports.isPlainObject = isPlainObject;
function flatMap(src, fn) {
    var res = [];
    src.forEach(function (x, i) { return res.push.apply(res, fn(x, i)); });
    return res;
}
exports.flatMap = flatMap;
