// @yext/search-core/dist/commonjs/src/models/searchservice/response/directanswer/BuiltInFieldType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuiltInFieldType = void 0;
/**
 * Possible built-in field types for {@link DirectAnswer.fieldType}.
 *
 * @public
 */
var BuiltInFieldType;
(function (BuiltInFieldType) {
    BuiltInFieldType["URL"] = "url";
    BuiltInFieldType["ComplexURL"] = "complex_url";
    BuiltInFieldType["IOSAppURL"] = "ios_app_url";
    BuiltInFieldType["AndroidAppURL"] = "android_app_url";
    BuiltInFieldType["FacebookURL"] = "facebook_url";
    BuiltInFieldType["Email"] = "email";
    BuiltInFieldType["InstagramHandle"] = "instagram_handle";
    BuiltInFieldType["TwitterHandle"] = "twitter_handle";
    BuiltInFieldType["Phone"] = "phone";
    BuiltInFieldType["Address"] = "address";
    BuiltInFieldType["Hours"] = "hours";
    BuiltInFieldType["Decimal"] = "decimal";
    BuiltInFieldType["Integer"] = "integer";
    BuiltInFieldType["SingleLineText"] = "single_line_text";
    BuiltInFieldType["RichText"] = "rich_text";
    BuiltInFieldType["MultiLineText"] = "multi_line_text";
})(BuiltInFieldType = exports.BuiltInFieldType || (exports.BuiltInFieldType = {}));
