// @yext/studio/lib/src/components/common/Modal.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import ReactModal from "react-modal";
import { ReactComponent as X } from "../../icons/x.svg";
import classNames from "classnames";
const customReactModalStyles = {
    content: {
        top: "50%",
        left: "50%",
        right: "50%",
        bottom: "auto",
        marginRight: "-95%",
        maxWidth: "600px",
        transform: "translate(-50%, -50%)",
    },
    overlay: {
        zIndex: 1000,
    },
};
export default function Modal({ isOpen, title, body, errorMessage, handleClose, handleConfirm, isConfirmButtonDisabled, confirmButtonText = "Ok", confirmButtonEnabledColor, }) {
    const footerClasses = classNames("mt-2 items-center", {
        "flex justify-between": errorMessage,
    });
    const confirmButtonClasses = classNames("ml-2 py-1 px-3 text-white rounded-md", {
        "bg-gray-400": isConfirmButtonDisabled,
        [confirmButtonEnabledColor ?? "bg-blue-600"]: !isConfirmButtonDisabled,
    });
    return (_jsxs(ReactModal, { isOpen: isOpen, onRequestClose: handleClose, style: customReactModalStyles, contentLabel: `${title} Modal`, ariaHideApp: false, shouldCloseOnEsc: true, shouldCloseOnOverlayClick: true, role: "dialog", children: [_jsxs("div", { className: "flex justify-between items-center font-bold mb-4", children: [title, _jsx("button", { onClick: handleClose, "aria-label": "Close Modal", children: _jsx(X, {}) })] }), body, _jsxs("div", { className: footerClasses, children: [errorMessage && (_jsx("div", { className: "flex justify-start text-red-600", children: errorMessage })), _jsxs("div", { className: "flex justify-end", children: [_jsx("button", { className: "ml-2", onClick: handleClose, children: "Cancel" }), _jsx("button", { className: confirmButtonClasses, disabled: isConfirmButtonDisabled, onClick: handleConfirm, children: confirmButtonText })] })] })] }));
}
