// @yext/pages/dist/dev/server/middleware/constants.js
import getPort, { portNumbers } from "get-port";
const devServerPort = await getPort({
  port: portNumbers(5173, 6e3)
});
const dynamicModeInfoText = `Dynamic mode enabled. Below are sample URLs generated based on the contents 
of the localData folder, but entity data will be re-fetched on each page load and reflect updates in real time.`;
const localDevUrlInfoText = `Local development urls will match production urls for all entity pages IF a slug 
is configured for each entity.`;
const disableProdUrl = "https://hitchhikers.yext.com/docs/pages/local-dev-urls";
const localDevUrlHelpText = `You can learn more/how to disable this <a href="${disableProdUrl}">here</a>.`;
const localModeInfoText = `Local mode enabled. Below are the URLs that are available based on 
the contents of the localData folder. Entity Data will only be refreshed upon regenerating the test data, so updates 
to entities will not be reflected in real time. To regenerate test data, enter the command yext sites 
generate-test-data. Also, entities other than the ones listed below will not be available unless run in dynamic mode.`;
const generateTestDataWarningText = `On server start-up, the call to regenerate test data failed. 
Check your network connection and that the CLI is authorized to the correct account. Functionality may be limited.`;
const noLocalDataErrorText = `No localData directory present, cannot generate example URLs. Run yext sites
generate-test-data to generate the localData directory.`;
export {
  devServerPort,
  dynamicModeInfoText,
  generateTestDataWarningText,
  localDevUrlHelpText,
  localDevUrlInfoText,
  localModeInfoText,
  noLocalDataErrorText
};
