// @yext/pages/dist/components/analytics/hooks.js
import { useContext } from "react";
import { getRuntime } from "../../util/index.js";
import { AnalyticsContext } from "./context.js";
import { concatScopes } from "./helpers.js";
import { useScope } from "./scope.js";
function useAnalytics() {
  const ctx = useContext(AnalyticsContext);
  if (!ctx) {
    return ctx;
  }
  if (getRuntime().name === "browser" && !window.setAnalyticsOptIn) {
    window.setAnalyticsOptIn = async () => {
      await ctx.optIn();
    };
  }
  const scope = useScope();
  return {
    trackClick(eventName, conversionData) {
      return ctx.trackClick(concatScopes(scope, eventName), conversionData);
    },
    setDebugEnabled(enabled) {
      return ctx.setDebugEnabled(enabled);
    },
    enableTrackingCookie() {
      return ctx.enableTrackingCookie();
    },
    identify(visitor) {
      return ctx.identify(visitor);
    },
    optIn() {
      return ctx.optIn();
    },
    pageView() {
      return ctx.pageView();
    },
    track(eventName, conversionData) {
      return ctx.track(concatScopes(scope, eventName), conversionData);
    }
  };
}
const useTrack = () => {
  return useAnalytics()?.track;
};
const usePageView = () => {
  return useAnalytics()?.pageView;
};
const useIdentify = () => {
  return useAnalytics()?.identify;
};
export {
  useAnalytics,
  useIdentify,
  usePageView,
  useTrack
};
